crypto/signals.c

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright 2016-2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "e_os.h"
+#include <string.h>
+
+#include <openssl/crypto.h>
+#include <internal/list.h>
+
+struct sigentry {
+    CRYPTO_SIGNAL p;
+    struct list list;
+};
+
+static struct list siglist;
+
+static ossl_inline int siglist_sigeq(struct list *l, void *data);
+
+static ossl_inline int siglist_sigeq(struct list *l, void *data)
+{
+    return container_of(l, struct sigentry, list)->p.signal == *(long*)data;
+}
+
+#if defined(OPENSSL_SYS_WINDOWS) || defined(_WIN32) || defined(__CYGWIN__)
+
+# include <windows.h>
+
+volatile PHANDLER_ROUTINE callback_handler = NULL;
+
+static BOOL WINAPI CRYPTO_SIGNAL_handler(DWORD dwType)
+{
+    if (list_empty(&siglist))
+        return FALSE;
+
+    struct list *l;
+    if ((l = list_find(&siglist, siglist_sigeq, (void*)&dwType)) == NULL)
+        return FALSE;
+
+    container_of(l, struct sigentry, list)->p.callback(dwType);
+    return TRUE;
+}
+
+static int CRYPTO_SIGNAL_arch_enable()
+{
+    callback_handler = (PHANDLER_ROUTINE) CRYPTO_SIGNAL_handler;
+    if (SetConsoleCtrlHandler(callback_handler, TRUE) == 0)
+        return 0;
+    return 1;
+}
+
+static int CRYPTO_SIGNAL_arch_disable()
+{
+    if (SetConsoleCtrlHandler(callback_handler, FALSE) == 0) {
+        callback_handler = NULL;
+        return 0;
+    }
+    return 1;
+}
+
+static int CRYPTO_SIGNAL_block_arch(CRYPTO_SIGNAL *p)
+{
+    return 1;
+}
+
+#elif defined(OPENSSL_SYS_UNIX) || defined (__unix__) || \
+      (defined (__APPLE__) && defined (__MACH__))
+
+# include <sys/types.h>
+# include <unistd.h>
+# include <signal.h>
+
+static int CRYPTO_SIGNAL_arch_enable(void)
+{
+    return 1;
+}
+
+static int CRYPTO_SIGNAL_arch_disable(void)
+{
+    return 1;
+}
+
+/**
+ * signal specifies which signal to (un)block/(un)mask
+ * callback is one of:
```

Comment:
should this decription of "callback" be up where the struct is defined?

Reply:
This is a POSIX-specific description, so I don't think having it in a generic context would help. Or we can get rid of it completely ...